### DISCUSSION TITLE
Using smart_open for reading gzip file with sftp

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [RaRe-Technologies/smart_open#827] [Q&A] [answered]
I want to read a remote gzip file line by line using sftp. I have a working solution using an sftp package but want to make it work with smart_open but the conversion is not working :(

The sftp code to convert is:
```
    with (
        open(remotepath, 'r', host, user, password, port, transport_params) as s,
        gzip.open(s.open(remotepath, mode="r"), "rt", encoding="utf-8") as g,
    ):
	while line := g.readline():
```
My attempt is:    
```
    from smart_open import open

    with open(remotepath, 'r', host, user, password, port, transport_params) as s:
        with gzip.open(s.open(remotepath, mode="r"), "rt", encoding="utf-8") as g:
            while line := g.readline():
```
Which generates the Traceback:
```
    with open(remotepath, 'r', host, user, password, port, transport_params) as s:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ...\smart_open\smart_open_lib.py", line 228, in open
    decoded = _encoding_wrapper(
              ^^^^^^^^^^^^^^^^^^
  ...\smart_open\smart_open_lib.py", line 436, in _encoding_wrapper
    fileobj = so_utils.TextIOWrapper(
              ^^^^^^^^^^^^^^^^^^^^^^^
ValueError: illegal newline value: 3373
```
What am I doing wrong?

**Comment by ddelange** [comment]
if your filename ends on `.gz`, smart_open will transparently decompress using gzip for you. if it does not, you can still force gzip (de)compression with the `compression` parameter:

https://github.com/piskvorky/smart_open?tab=readme-ov-file#compression-handling

```pycon
>>> from smart_open import open
>>> with open('sftp://smart_open/tests/test_data/1984.txt.gzip', compression='.gz') as fin:
...     print(fin.read(32))
It was a bright cold day in Apri
```

**Reply by dineshbvadhia** [reply]
The remote file ends with a `.gz`. So, do you mean open like? :
```
    with open(remotepath, 'r', host, user, password, port, transport_params) as g:
        while line := g.readline():
```

**Reply by ddelange** [reply]
yes exactly, that will first decompress and then yield lines

**Reply by dineshbvadhia** [reply]
This works :) :
`    with open(remotepath, 'r', transport_params) as g:
`

but not this:
`    with open(remotepath, 'r', host, user, password, port, transport_params) as g:
`

which gives the same Traceback ie:
```
    with open(remotepath, 'r', host, user, password, port, transport_params) as g:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  ...\smart_open\smart_open_lib.py", line 228, in open
    decoded = _encoding_wrapper(
              ^^^^^^^^^^^^^^^^^^
  ...\smart_open\smart_open_lib.py", line 436, in _encoding_wrapper
    fileobj = so_utils.TextIOWrapper(
              ^^^^^^^^^^^^^^^^^^^^^^^
ValueError: illegal newline value: 3373
```
The port number is 3373 which is passed to smart_open as a str.

transport_params is a dataclass:

`TransportParams(connect_kwargs=ConnectKwargs(host='127.0.0.1', user='admin', password='admin', port='3373'), keyfile=None, passphrase=None)
`

**Reply by ddelange** [reply]
try using keyword arguments instead of positional arguments. it looks like your variable `port` is now being passed into the `newline` argument

**Reply by dineshbvadhia** [reply]
`with open(remotepath, mode='r', host=host, user=user, password=password, port=port, transport_params=transport_params) as g:`

generates traceback:
```
with open(remotepath, mode='r', host=host, user=user, password=password, port=port, transport_params=transport_params) as g:\n         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n
TypeError: open() got an unexpected keyword argument 'host'"
```

This is weird :(

**Reply by ddelange** [reply]
i think passing a dict to transport_params should be enough?

**Reply by ddelange** [reply]
aha, i see. if you're making use of the top level `smart_open.open` (and not `smart_open.ssh.open`), host and port etc should be embedded in the uri.

top level open will use `smart_open.ssh.open_uri`: https://github.com/piskvorky/smart_open/blob/5a826134eb23c86087d06db3b78b25484e7e2e6d/smart_open/smart_open_lib.py#L366

which in turn parses host and port etc out of the uri, and propagates to `smart_open.ssh.open`: https://github.com/piskvorky/smart_open/blob/5a826134eb23c86087d06db3b78b25484e7e2e6d/smart_open/ssh.py#L113

**Reply by dineshbvadhia** [reply]
I'm using a cli to kickoff a program which calls smart_open eg:

`python myprogram.py ... --remotepath test.gz --host 127.0.0.1 --user admin --password admin --port 3373`

Is `smart_open.open` or `smart_open.ssh.open` appropriate?

**Reply by mpenkov** [reply]
Please see the documentation.

```python
help(smart_open)
```

smart_open.open is the function that you want.

**Reply by ddelange** [reply, marked as answer]
```pycon
>>> from smart_open import open
>>> with open('sftp://username:password@host:port/path/file.gz') as fin:
...     print(fin.read(32))
It was a bright cold day in Apri
```

**Reply by ddelange** [reply]
You can also use an SSH key file instead of embedding a password in the uri: https://github.com/piskvorky/smart_open/pull/522/files

**Reply by dineshbvadhia** [reply]
Ok, but still having a problem. As before, this works:
```
    from smart_open import open
    
    with open(remotepath, 'r', transport_params) as g:
```
remotepath is a fully qualified posix `Path`:

This doesn't work:
`    with open(uri, 'r', transport_params) as g:`

I'm developing on Windows 11 Pro (with latest updates). Does smart_open expect a Windows path or uri in a specific format and/or pre-processed?

I'm happy to use the working solution above but it would be good to identify why the uri method doesn't work. It could just be Windows hell :(

**Reply by ddelange** [reply]
the uri should be formatted as in my last code snippet above. that will trigger the uri parsing mechanic as described in my comment [before](https://github.com/piskvorky/smart_open/discussions/827#discussioncomment-9949181)

**Reply by dineshbvadhia** [reply]
Ok, it works. The path wasn't prescribed relative to where the sftpserver was running. Phew. Thank-you for your patience :)

About the use of an SSH key file: My software is for general use by developers. Wrt sftp, the user name and password are mandatory, a keyfile and passphrase are optional. Currently, the setup is a CLI as shown earlier https://github.com/piskvorky/smart_open/discussions/827#discussioncomment-9949370. So, what is a secure way to start the program which includes sftp and non-sftp arguments?

**Reply by ddelange** [reply]
I generally dislike passing passwords in via CLI, as history of CLI commands is often saved in plaintext. You can make `——password` optional (I guess you're using [click](https://pypi.org/project/click/)?) and add another optional `--key-filename`, so a developer can choose between providing his password and providing the path to his private key (which should be trusted by the remote sshd).

**Reply by dineshbvadhia** [reply]
Sounds good. Again, thanks. Will close the discussion :)
